prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,208 @@
+*/
+
+// Package projectqueries is a plugin to auto add pull requests to project boards based on a search query
+package projectqueries
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName = "project-queries"
+)
+
+// TODO Create a new handler for issues, look in hook/server.go
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The projects plugin auto adds issues/PRs to projects based on a search query",
+		Config: func(config *plugins.Configuration) map[string]string {
+			configMap := make(map[string]string)
+			configString := ""
+			for _, projectQuery := range config.ProjectQueries {
+				projectString := fmt.Sprintf("Project Org: \"%s\", Project Name: \"%s\", Project Column: \"%s\"", projectQuery.Org, projectQuery.Name, projectQuery.Column)
+				labelsString := ""
+				for _, label := range projectQuery.Query.Labels {
+					labelsString = fmt.Sprintf("%s label:%s", labelsString, label)
+				}
+				queryString := fmt.Sprintf("Query: \"org:%s state:%s%s\"", projectQuery.Query.Org, projectQuery.Query.State, labelsString)
+				configString = fmt.Sprintf("%s\n%s %s", configString, projectString, queryString)
+			}
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectQueries, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if len(projectQueries) == 0 {
+		return nil
+	}
+
+	// Only respond to label add or issue/PR open events
+	if !isPRChanged(pe) {
+		return nil
+	}
+
+	// Get any ProjectQueries that match this PR
+	matchedProjectQueries, err := getMatchingProjectQueries(gc, projectQueries, pe)
+	if err != nil {
+		return err
+	}
+	if len(matchedProjectQueries) == 0 {
+		return nil
+	}
+
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		le.WithError(err)
+	}
+
+	return nil
+}
+
+// labelsInLabels check that all of the queryLabels (the first list), are present in
+// the list of github.Labels (the second list)
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
```

Comment:
You could use:

https://github.com/kubernetes/test-infra/blob/2f2f372c620afd5c000f5068d89db21f3865ddf6/prow/github/helpers.go#L26-L34

Or add your own there. If adding your own, consider using `sets.String.HasAll()`

Reply:
I have added a `HasLables` function to `helpers.go`, changed the `project_queries` plugin to use the helper function and removed the redundant `labelsInLabels` function.

The new `HasLabels` function iterates through the list of strings, making use of the existing `HasLabel` function for each rather than using `sets.String.HasAll()`. This is to ensure consistent behaviour as `HasLabel` compares both strings converted to lower case.